lib/transfer.c

```diff
@@ -140,15 +179,63 @@ CURLcode Curl_fillreadbuffer(struct connectdata *conn, size_t bytes,
   }
 #endif
 
-  if(data->req.upload_chunky) {
+#ifndef CURL_DISABLE_HTTP
+  if(data->state.trailers_state == HTTP_TRAILERS_INITIALIZED) {
+    /* at this point we already verified that the callback exists
+       so we compile and store the trailers buffer, then proceed */
+    infof(data,
+          "Moving trailers state machine from initialized to sending.\n");
+    data->state.trailers_state = HTTP_TRAILERS_SENDING;
+    data->state.trailers_buf = Curl_add_buffer_init();
+    if(!data->state.trailers_buf) {
+      failf(data, "Unable to allocate trailing headers buffer !");
+      return CURLE_OUT_OF_MEMORY;
+    }
+    data->state.trailers_bytes_sent = 0;
+    Curl_set_in_callback(data, true);
+    data->set.trailer_callback(&trailers, data->set.trailer_data);
+    Curl_set_in_callback(data, false);
+    c = Curl_http_compile_trailers(trailers, data->state.trailers_buf, data);
+    if(c != CURLE_OK) {
+      Curl_add_buffer_free(&data->state.trailers_buf);
+      curl_slist_free_all(trailers);
+      return c;
+    }
+    infof(data, "Successfully compiled trailers.\r\n");
+    curl_slist_free_all(trailers);
+  }
+#endif
+
+  /* if we are transmitting trailing data, we don't need to write
+     a chunk size so we skip this */
+  if(data->req.upload_chunky
+     && data->state.trailers_state == HTTP_TRAILERS_NONE) {
```

Comment:
When splitting long expressions, we usually prefer to split the lines with the operator on the previous line, and I would recommend parentheses:

    (data->req.upload_chunky &&
     (data->state.trailers_state == HTTP_TRAILERS_NONE))